discovery/dns/dns_test.go

```diff
@@ -0,0 +1,177 @@
+				Port:            80,
+				Type:            "AAAA",
+			},
+			lookup: func(name string, qtype uint16, logger log.Logger) (*dns.Msg, error) {
+				return &dns.Msg{
+						Answer: []dns.RR{
+							&dns.AAAA{AAAA: net.IPv6loopback},
+						},
+					},
+					nil
+			},
+			expected: []*targetgroup.Group{
+				&targetgroup.Group{
+					Source: "web.example.com.",
+					Targets: []model.LabelSet{
+						{"__address__": "[::1]:80", "__meta_dns_name": "web.example.com."},
+					},
+				},
+			},
+		},
+		{
+			name: "SRV record query without trimming dot",
+			config: SDConfig{
+				Names:           []string{"_mysql._tcp.db.example.com."},
+				RefreshInterval: model.Duration(time.Minute),
+				TrimFinalDot:    false,
+			},
+			lookup: func(name string, qtype uint16, logger log.Logger) (*dns.Msg, error) {
+				return &dns.Msg{
+						Answer: []dns.RR{
+							&dns.SRV{Port: 3306, Target: "db1.example.com."},
+							&dns.SRV{Port: 3306, Target: "db2.example.com."},
+						},
+					},
+					nil
+			},
+			expected: []*targetgroup.Group{
+				&targetgroup.Group{
+					Source: "_mysql._tcp.db.example.com.",
+					Targets: []model.LabelSet{
+						{"__address__": "db1.example.com.:3306", "__meta_dns_name": "_mysql._tcp.db.example.com."},
+						{"__address__": "db2.example.com.:3306", "__meta_dns_name": "_mysql._tcp.db.example.com."},
+					},
+				},
+			},
+		},
+		{
+			name: "SRV record query with trimming dot",
+			config: SDConfig{
+				Names:           []string{"_mysql._tcp.db.example.com."},
+				RefreshInterval: model.Duration(time.Minute),
+				TrimFinalDot:    true,
+			},
+			lookup: func(name string, qtype uint16, logger log.Logger) (*dns.Msg, error) {
+				return &dns.Msg{
+						Answer: []dns.RR{
+							&dns.SRV{Port: 3306, Target: "db1.example.com."},
+							&dns.SRV{Port: 3306, Target: "db2.example.com."},
+						},
+					},
+					nil
+			},
+			expected: []*targetgroup.Group{
+				&targetgroup.Group{
+					Source: "_mysql._tcp.db.example.com.",
+					Targets: []model.LabelSet{
+						{"__address__": "db1.example.com:3306", "__meta_dns_name": "_mysql._tcp.db.example.com."},
+						{"__address__": "db2.example.com:3306", "__meta_dns_name": "_mysql._tcp.db.example.com."},
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			sd := NewDiscovery(tc.config, nil)
+			sd.lookupFn = tc.lookup
+
+			ch := make(chan []*targetgroup.Group, 1)
+			ctx, cancel := context.WithTimeout(context.Background(), 200*time.Millisecond)
+			sd.Run(ctx, ch)
+			cancel()
+
+			var tg []*targetgroup.Group
+			select {
+			case tg = <-ch:
+			default:
+				if tc.expected == nil {
+					return
+				}
+				t.Fatal("Empty channel")
+			}
+
+			if len(tg) != len(tc.expected) {
+				t.Fatalf("Invalid number of target groups, exp: %d, got: %d", len(tc.expected), len(tg))
+			}
+			for i := range tg {
+				if !tg[i].Equal(tc.expected[i]) {
```

Comment:
In many other places in Prometheus we either use https://github.com/prometheus/prometheus/blob/3b5dea4e6db822533a2e3d2eb1d253ec04853493/util/testutil/testing.go#L64-L70 or `testify`'s `require.Equal()`, like here: https://github.com/prometheus/prometheus/blob/2bd510a63e48ac6bf4971d62199bdb1045c93f1a/storage/fanout_test.go#L36

The benefit of the latter is nice built-in diff formatting when something isn't equal. Can we just use that here? Then we also wouldn't need the new `Equal()` method on the target group.

Reply:
Sure I'll update the PR.